lib/rubocop/cop/rails/orm_abstraction.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # Avoid breaking ORM abstraction.
+      # @example
+      #
+      #   # bad
+      #   where(foo_id: foo.id)
+      #
+      #   # bad
+      #   create(foo_id: foo.id)
+      #
+      #   # good
+      #   where(foo: foo)
```

Comment:
I'm sorry to cause similar questions.

What is the aim of this cop?
I'm not sure the reason why `model` is better than `model.id` as argument value.

The following is not `AR::Base.where` shown in the example...
For example, if the `AR::Base.find` method is used without calling `id`, it will be warned as follows.

```console
> Model.find(model)
ArgumentError: You are passing an instance of ActiveRecord::Base to `find`. Please pass the id of the object by calling `.id`.
> Model.find(model.id) # Not warned
```

https://github.com/rails/rails/blob/v5.2.0/activerecord/lib/active_record/relation/finder_methods.rb#L444-L449

In `AR::Base.where` and `AR::Base.find_by` the above is not warned, but there seems to be no reason to omit the `id`. 🤔 

```ruby
Model.where(id: model)   # Not warned
Model.find_by(id: model) # Not warned
```

Reply:
> I'm not sure the reason why model is better than model.id as argument value.

Referencing IDs while working with objects breaks the ORM abstraction
because you are using database concerns. By avoiding that when possible
you can avoid bugs and hard to reason about code.

So, if all you have is an id, then by all means, use the id. If you have an object, it is better to use the object.

> For example, if the AR::Base.find method is used without calling id, it will be warned as follows.

This doesn't affect `.find` or derivatives. Maybe the scope of this can be reduced to `.create` and `.new` so that it doesn't affect DB query methods? I included `where` because `where` supports using models and that means you can save like 6 characters.

>Model.where(id: model)   # Not warned
Model.find_by(id: model) # Not warned

I am not sure why you would do this over `model.reload`.

Reply:
Thanks for the replying.

I'm wondering about the difference in what I accept in search interfaces `find`, `where`, and `find_by` in the current AR.

`find` method argument requires `model.id`. This is a change made in Rails 4.2. Until then, I could call it with `model`.
https://github.com/rails/rails/commit/d35f0033c7dec2b8d8b52058fb8db495d49596f7

As of now there are no similar changes in other `where`, `find_by`. However, it is doubtful how much benefit a user has about using a different interface (`model`) than `find`.

So, at least for this example code I doubt that it is designed to urge users to omit `id`.
I think that it is doubtful to correct the omission of `id` by default for users.

Reply:
`find` takes a single argument. `create`/`where`/`new` take keyword arguments. Keyword argument ActiveRecord methods tend to have special code to take advantage of passing a model. They are completely different and don't really have any similarities to `find` beyond both being used to build sql queries.

Reply:
As a practical problem, I have other concern.

There is a case where `model.pluck` is used to acquire a foreign key and use it as an argument of `model.where`. It is a technique to avoid wasting memory by instantiating an AR object.

While preventing the intention of using such performance technique, it seems that there is no advantage over that.

Reply:
>There is a case where model.pluck is used to acquire a foreign key and use it as an argument of model.where. It is a technique to avoid wasting memory by instantiating an AR object.

You don't need to do that. Rails will build the query for you if you pass in another relation:

```
[1] pry(main)> User.where(posts: Post.where(title: "foo"))
  User Load (0.3ms)  SELECT "users".* FROM "users" WHERE "users"."id" IN (SELECT "posts"."user_id" FROM "posts" WHERE "posts"."title" = ?)  [["title", "foo"]]
```

Reply:
In such a simple case it may work.

However, when composing a complicated combination of queries, it is not necessarily preferable that inline code is preferable.

I don't think it's better to warn uniformly about such a case.
This is because adding restrictions on how to build query can compromise the best way to describe queries for users.

So, building a query for business is complicated, I think it is difficult for me to force this style there.

Reply:
> So, building a query for business is complicated

I don't appreciate being attacked, ad hominem.

I get that you are busy and don't have time to really understand what a new cop is trying to do, but maybe next time you don't understand what a pull request is about, ask the author for clarification _**before**_ attempting to tell them the reasons their patch wont work. You come across as hostile and then I have to tell you that you didn't understand the patch which is going to annoy you.

> This is because adding restrictions on how to build query can compromise the best way to describe queries for users.

I have encountered this before. I am going to need an example to be able to come up with an alternate solution.